server/proxy/synchronize_test.go

```diff
@@ -0,0 +1,212 @@
+				"id1": {
+					Manifest: apps.Manifest{
+						AppID:            "id1",
+						AppType:          "http",
+						Version:          "1",
+						OnVersionChanged: &apps.Call{},
+					},
+					Disabled: false,
+				},
+			},
+			inputManifests: map[apps.AppID]*apps.Manifest{
+				"id2": {
+					AppID:   "id2",
+					AppType: "http",
+				},
+			},
+			needUpstream: false,
+
+			expectedApps: map[apps.AppID]*apps.App{
+				"id1": {
+					Manifest: apps.Manifest{
+						AppID:            "id1",
+						AppType:          "http",
+						Version:          "1",
+						OnVersionChanged: &apps.Call{},
+					},
+					Disabled: false,
+				},
+			},
+			expectedManifests: map[apps.AppID]*apps.Manifest{
+				"id2": {
+					AppID:   "id2",
+					AppType: "http",
+				},
+			},
+		},
+		{
+			name: "update version",
+			inputApps: map[apps.AppID]*apps.App{
+				"id1": {
+					Manifest: apps.Manifest{
+						AppID:            "id1",
+						AppType:          "http",
+						Version:          "1",
+						OnVersionChanged: &apps.Call{},
+					},
+					Disabled: false,
+				},
+			},
+			inputManifests: map[apps.AppID]*apps.Manifest{
+				"id1": {
+					AppID:            "id1",
+					AppType:          "http",
+					Version:          "2",
+					OnVersionChanged: &apps.Call{},
+				},
+			},
+			needUpstream: true,
+
+			expectedApps: map[apps.AppID]*apps.App{
+				"id1": {
+					Manifest: apps.Manifest{
+						AppID:            "id1",
+						AppType:          "http",
+						Version:          "2",
+						OnVersionChanged: &apps.Call{},
+					},
+					Disabled: false,
+				},
+			},
+			expectedManifests: map[apps.AppID]*apps.Manifest{
+				"id1": {
+					AppID:            "id1",
+					AppType:          "http",
+					Version:          "2",
+					OnVersionChanged: &apps.Call{},
+				},
+			},
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			p := createProxy(tc.inputApps, tc.inputManifests)
+
+			if tc.needUpstream {
+				ctrl := gomock.NewController(t)
+				defer ctrl.Finish()
+
+				cr := &apps.CallResponse{
+					Type: apps.CallResponseTypeOK,
+					Data: "some data",
+				}
+				bb, _ := json.Marshal(cr)
+				reader := ioutil.NopCloser(bytes.NewReader(bb))
+				up := mock_upstream.NewMockUpstream(ctrl)
+				up.EXPECT().Roundtrip(gomock.Any(), gomock.Any()).Return(reader, nil)
+
+				upDetector := mock_detector.NewMockDetector(ctrl)
+				upDetector.EXPECT().UpstreamForApp(gomock.Any()).Return(up, nil)
+
+				p.upstreamDetector = upDetector
+			}
+
+			err := p.SynchronizeInstalledApps()
+			require.NoError(t, err)
+			installedApps := p.store.App.AsMap()
+			require.Equal(t, tc.expectedApps, installedApps)
+			listedApps := p.store.Manifest.AsMap()
+			require.Equal(t, tc.expectedManifests, listedApps)
+			for _, app := range installedApps {
+				if _, isListed := listedApps[app.AppID]; !isListed {
+					require.False(t, p.AppIsEnabled(app))
+				}
+			}
+		})
+	}
+}
+
+func createProxy(a map[apps.AppID]*apps.App, m map[apps.AppID]*apps.Manifest) *Proxy {
```

Comment:
Can we name this `createTestProxy`, so it's obvious to callers that it's creating a `Proxy` for testing?